flink-libraries/flink-table/src/main/java/org/apache/flink/table/plan/logical/rel/util/WindowAggregateUtil.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.plan.logical.rel.util;
+
+import java.io.Serializable;
+import java.util.List;
+
+import org.apache.calcite.rel.core.Window.Group;
+import org.apache.calcite.rel.logical.LogicalWindow;
+import org.apache.calcite.rex.RexLiteral;
+
+import com.google.common.collect.ImmutableList;
+
+public class WindowAggregateUtil implements Serializable {
+
+	private static final long serialVersionUID = -3916551736243544540L;
+
+	private LogicalWindow windowPointer = null;
+
+	public WindowAggregateUtil() {
+
+	}
+
+	public WindowAggregateUtil(LogicalWindow window) {
+		this.windowPointer = window;
+
+	}
+
+	/**
+	 * A utility function that checks whether a window is partitioned or it is a
+	 * global window.
+	 * 
+	 * @param LogicalWindow
+	 *            window to be checked for partitions
+	 * @return true if partition keys are defined, false otherwise.
+	 */
+	public boolean isStreamPartitioned(LogicalWindow window) {
+		// if it exists a group bounded by keys, the it is
+		// a partitioned window
+		for (Group group : window.groups) {
+			if (!group.keys.isEmpty()) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	public int[] getKeysAsArray(Group group) {
+		if (group == null) {
+			return null;
+		}
+		return group.keys.toArray();
+	}
+
+	/**
+	 * This method returns the [[int]] lowerbound of a window when expressed
+	 * with an integer e.g. ... ROWS BETWEEN [[value]] PRECEDING AND CURRENT ROW
+	 * 
+	 * @param constants
+	 *            the list of constant to get the offset value
+	 * @return return the value of the lowerbound if available -1 otherwise
+	 */
+
+	public int getLowerBoundary(ImmutableList<RexLiteral> constants) {
+		return ((Long)constants.get(0).getValue2()).intValue();
```

Comment:
I think maybe we can not get the boundary just constants[0], such as 
```
select sum(2) over (order by rowtime() range interval '1' hour preceding) as u3 FROM  OrderA
```
the boundary will be constants[1],
you can use constants.get(lowerBound.getOffset.asInstanceOf[RexInputRef].getIndex - inputRowType.getFieldCount)